### DISCUSSION TITLE
use Ghosts Shadows

### DISCUSSION THREAD

**Discussion by redwiz666** [cmu-sei/GHOSTS#431] [Q&A]
I am able to setup ghosts shadows and can successfully login using the UI on port 7860. I have ran a few queries to make sure the LLM is working correctly, but how do we make ghost clients use it? I looked in application.json and there is no config place for it. Mainly I am wanting to use the web_content and excel_content models.

**Comment by cvanslyke** [comment]
I've got a similar issue, but with Pidgin. How do I get the client to use the shadows chat API for content?

**Comment by nblair2** [comment]
Also interested in this

**Comment by sei-dupdyke** [comment]
This is still under development (sorry, I am swamped with other projects currently) — you'd have to script it yourself for anything outside some of the basic use cases that exist (chat and social media posting) done from animation jobs on the API server.

Basically, the thinking is that this sort of thing is done at the API server level and then sent down to the clients to execute, in order to centralize calls to the LLM, but also to reduce overhead for things where the agent has to know about other agents and what they know. So you won't see any config for Shadows or an LLM at the client level. Another primary driver for this is we didn't want clients making a bunch of requests off to an array of services, because in an exercise we may want to hide or whitelist these. With everything from the client going through the API, that is much easier to manage.

I haven't used Pidgin much, but I assume it has the same concerns as doing chat over MatterMost.

Web content really is early, but its for things like a client requesting a page that we dynamically generate on the fly. Web scraping for content to move in a gapped range is hard, this was some early work to see if we could skip that and dynamically load/create it on the fly.

**Reply by nblair2** [reply]
Got it - thanks for the detailed answer!

I'll keep an eye out for whenever the API gets updated to include ollama integration. Completely agree on the centralized server architecture for all of the reasons you listed.

As for web content - I took a look at pandora and really like the idea. Effectively models any site where you just need raw traffic without a need for specific content. I think the next step would be adding self-referencing links on pages, so if you go to `https://pandora.com/some/page.html`, it can also insert a couple of links to `https://pandora.com/other/page.html` and `https://pandora.com/some/picture.png`.

**Reply by sei-dupdyke** [reply]
Is there a particular thing you all would want to see first? I can prioritize that use case (I'm not confident my use cases are the best for the broadest array of people).

Thanks!

**Reply by nblair2** [reply]
I think driving agent content (email replies, office document data, etc) via ollama is first in my mind.

**Reply by dh4340** [reply]
> Got it - thanks for the detailed answer!
> 
> I'll keep an eye out for whenever the API gets updated to include ollama integration. Completely agree on the centralized server architecture for all of the reasons you listed.
> 
> As for web content - I took a look at pandora and really like the idea. Effectively models any site where you just need raw traffic without a need for specific content. I think the next step would be adding self-referencing links on pages, so if you go to `https://pandora.com/some/page.html`, it can also insert a couple of links to `https://pandora.com/other/page.html` and `https://pandora.com/some/picture.png`.

Thank you for your appreciation of the enhancements to Pandora.
Constructing realistic websites with a compact LLM within the requisite timeframe to embed links is rather challenging.
When I integrated Ollama into Pandora, it was to build upon an excellent concept. For embedded links, we would require an agent-style development interface.

I concur with @nblair2 on the necessity for agents to employ shadows for content, exchanging emails among themselves and making decisions. A promising way forward might be to utilise agentic AI.
